### ISSUE TITLE
Exiting the container does not trigger a `runtime delete` from Docker

### ISSUE THREAD

**Issue by sboeuf** [moby/moby#38233]
**Description**

Kata Containers was working great with Docker `18.06`, but when moved to `18.09` I realized that our container (VM in our case) was not properly destroyed. This being tied to the fact that when the container process terminates (let's say `exit` from the shell), we expect `docker` (and `containerd`) to call into `runtime delete <container-id>` after it detected the container process terminated.

Apparently the logic seems to have changed because the behavior is now different and docker does not try to `delete` my container.

**Comment by sboeuf** [comment]
I think I had this issue because `docker-containerd` has been renamed into `containerd`...

From the release note:
- Remove 'docker-' prefix for containerd and runc binaries docker/engine#61 / moby/moby#37907, docker-ce-packaging#241

Problem is that people have `containerd` already installed on their system somewhere, with the `containerd` service already configured. And unfortunately this might not be the one installed by your update of Docker.

This means that when docker service is started, the containerd service is also started. Unfortunately, it's not using the right binary for those cases where the containerd service did exist before.

**Comment by sboeuf** [comment]
Weird... I thought it was fixed cause it works on my machine locally but on some other machines (VMs) I hit the issue I was hitting earlier:
```
dockerd[32303]: time="2018-11-21T01:39:01.908771921Z" level=error msg="d312e3aae3da1a7b0cb846aaffd4fd8b7a662efcf5304d0827c19a2344087de9 cleanup: failed to delete container from containerd: cannot delete running task d312e3aae3da1a7b0cb846aaffd4fd8b7a662efcf5304d0827c19a2344087de9: failed precondition"
```

If anyone (@amshinde @chavafg @devimc @jodh-intel @egernst) wants to investigate, here is the code path:
 [Cleanup()](https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/daemon/start.go#L251-L253) ==> [Delete()](https://github.com/containerd/containerd/blob/efb04a3361842b3e33d73e32e6d8d90293731169/container.go#L143-L146) ==> [loadTask()](https://github.com/containerd/containerd/blob/efb04a3361842b3e33d73e32e6d8d90293731169/container.go#L275-L299)

For some reason, `loadTask()` is failing. But I haven't really investigated further because when I wanted to debug, I built containerd from sources and it worked for me. But now I realize it doesn't apply for every use case...

**Comment by jodh-intel** [comment]
I've recreated the issue and put tabular results here:

- https://github.com/kata-containers/runtime/issues/926#issuecomment-440675909

Testing with 18.09 and `runc` if I create a container and then exit by typing `exit`, I see the following calls to `/usr/sbin/runc`:

```
create
state
start
state
state
delete
```

Note: no call to `kill` [which seems unexpected](https://github.com/opencontainers/runtime-spec/blob/master/runtime.md#lifecycle).

But if I test with `runc` using 18.06.1, the same container lifecycle generates the following calls to `/usr/bin/docker-runc`:

```
create
state
start
state
kill
state
kill
delete
```

So the behaviour has changed.

**Comment by thaJeztah** [comment]
Wonder if this is related to https://github.com/containerd/containerd/pull/2773 / https://github.com/containerd/containerd/issues/2744

**Comment by jodh-intel** [comment]
Using docker 18.09 with containerd v1.1.2 I see the expected behaviour with both [`kata-runtime`](https://github.com/kata-containers/runtime/) and `runc`.

With 18.09 and and containerd v1.2.0 and `kata-runtime` I see the following if the user types `exit` to kill the workload:

```
create
state
start
state
state
state
```

To summarise, with docker 18.09 and containerd 1.2.0:

- `runc`: `delete` is being called, but `kill` is not.
-  `kata-runtime`: neither `delete` or `kill` is being called.

So, it appears that containerd changed its behaviour between 1.1.2 and 1.2.0. Further at 1.2.0 it is making different calls, depending on the runtime being used. This seems to confirm @sboeuf's suspicions on https://github.com/kata-containers/runtime/issues/926 that containerd is not noticing when our [`kata-shim`](https://github.com/kata-containers/shim) process (which represents the workload) dies.

**Comment by jodh-intel** [comment]
I've bisected and found the culprit - see https://github.com/containerd/containerd/issues/2824.

@crosbymichael - The problem is that the commit above appears to break the OCI behavioural contract between the management layer and the runtime. I'm not clear how `runc` tolerates this, but since two runtime calls seem to have been optimised away in the "user-initiated workload shutdown" scenario, there is now no way for Kata (and I suspect other non-`runc` runtimes) to shutdown cleanly in that scenario. Hence, a pretty serious containerd bug imho.

Until we understand how this issue can be fixed, Kata users will have to stay on Docker 18.06.1 with the bundled containerd v1.1.2 fwics.

**Comment by sboeuf** [comment]
@crosbymichael 
>Does it matter that there are kill calls after the process exits? It's part of the kill all logic.

I'm not entirely sure about the issue here, but I can confirm what @jodh-intel found. If I re-enable the `kill --all` for any container, Kata works as expected.
I don't understand why `kill --all` has been disabled only for specific cases related to the presence of a PID namespace, but something is going wrong because now the containerd-shim does not trigger the `delete` that is needed... Need more investigation about why `containerd-shim` does not trigger `delete`. The `kill --all` itself is not strictly needed, but I don't get where/why `containerd-shim` waits.

**Comment by lifubang** [comment]
I think moby will trigger `delete` when he received `TaskExit` event with the ID is the container's ID. Please see:
https://github.com/containerd/containerd/blob/master/runtime/v1/shim/service.go#L525-L533
I have tested moby+containerd(1.1/1.2)+runc，it works fine.

I guess maybe the kataruntime don't exit when sh exit. Because containerd-shim needs `unix.SIGCHLD` signal to trigger event `TaskExit` in `checkProcess`. Please check.

**Comment by sboeuf** [comment]
@lifubang thanks for the pointers, that's exactly where I was trying to put some debug traces. Talking about traces, any chance you know how I can simply add more logs to the code?
I've added things like this:
```go
log.G(s.context).Info("checkProcesses() 1")
```

but even enabling --log-level=debug won't show me any traces from the `journal`.

**Comment by lifubang** [comment]
dockerd --debug

**Comment by sboeuf** [comment]
@thaJeztah @crosbymichael @lifubang 
Hi folks, I'm sorry to use this Github issue for this purpose, but I don't have your email addresses and I didn't know hot to contact you.

This issue raised a good point on our side, which is the fact that we'd like to make sure we don't discover too late that Kata Containers is not behaving as expected with latest Docker version.
In order to prevent from that, we thought it could be nice to integrate Kata Containers to the Docker CI if that is possible.
There's also the worry to have devicemapper storage driver to be fully removed at some point, and we'd also like to make sure we have a proper backup/solution to anticipate that.

In short, it'd be great if we could have a chat with you (and potentially other stakeholders from containerd and Docker) to make sure we can find some solutions :)

Thanks!

**Comment by thaJeztah** [comment]
@sboeuf feel free to send me an e-mail at `sebastiaan <curly-thing> docker com`, and I can add some people to that thread 👍

**Comment by sboeuf** [comment]
@thaJeztah will do! Thanks!

**Comment by sboeuf** [comment]
https://github.com/kata-containers/runtime/issues/926 fixed the issue, and this can be now closed!
